Actions stable release

bholmesdev opened the pull request:
## Changes

This makes Actions stable for v4.15.
- Remove `experimental` flag 
- Remove deprecated `getActionProps()` API
- Add check for whether actions are used before mounting routes and middleware
- Add check for a define action symbol before registering an action function. Maybe overkill, but this 100% ensures the change is non-breaking for users that _happen_ to have an `actions/index.ts` file that isn't actions related.

## Testing

Remove experimental flag and `getActionProps()` from tests

## Docs

https://github.com/withastro/docs/pull/8980

bholmesdev commented on packages/astro/src/actions/index.ts:
```diff
@@ -20,10 +20,17 @@ export default function astroActions({
 		name: VIRTUAL_MODULE_ID,
 		hooks: {
 			async 'astro:config:setup'(params) {
+				isActionsUsed = actionsFileExists(fs, params.config.srcDir);
```
It's not an integration. It's in core 🙃

bholmesdev replied:
I have realized this check isn't full-proof though. To make it backwards compatible, we should check if the `server` export is present before booting up actions

florian-lefebvre replied:
I know it's in core but we are actually inserting an integration in the astro config, that's what i meant. like instead of checking in each hook, check as early as possible

bholmesdev replied:
Ah, perhaps we could. The issue is that `astro:server:setup` hook, which we _do_ want to mount to detect when actions are added in your project. If we conditionally add the whole integration, we can't check for this

florian-lefebvre replied:
Ah I didn't know, current approach seems go then

bholmesdev commented on packages/astro/src/integrations/hooks.ts:
```diff
@@ -130,10 +130,8 @@ export async function runHookConfigSetup({
-		const { default: actionsIntegration } = await import('../actions/index.js');
-		settings.config.integrations.push(actionsIntegration({ fs, settings }));
-	}
+	const { default: actionsIntegration } = await import('../actions/index.js');
```
Good point! Let's move to a top-level import

bholmesdev commented:
@florian-lefebvre I revisited your comment on simplifying the integration. I realized we could do so! I split things up so we have `vitePluginActions()` to define the `astro:actions`, module, and the `astroIntegrationActionsRouterHandler()` to _only_ inject routes when actions are used. I moved that file watcher to a vite plugin. Should solve a failing test in the process too 👍

sarah11918 commented on .changeset/spicy-suits-explode.md:
```diff
@@ -0,0 +1,23 @@
+
+The Astro Actions API introduced behind a flag in [v4.8.0](https://github.com/withastro/astro/blob/main/packages/astro/CHANGELOG.md#480) is no longer experimental and is available for general use.
+
+Astro Actions allow you to define and call backend functions with type-safety. Actions perform data fetching, JSON parsing, and input validation for you. This can greatly reduce the amount of boilerplate needed compared to using an API endpoint.
```
No notes on the included text! Just giving you permission in case you wanted to expand here a bit with an example, maybe like how the experimental changeset would have looked (updated and correct of course) or copying from the deleted experimental config reference content below.

I will not un-approve if this gets longer! At your discretion, @bholmesdev !

bholmesdev replied:
Haha, appreciate that. I'll copy a snippet from the 4.15 release blog post

sarah11918 reviewed (approved):
Just giving Ben more permission to hype it up in the changeset as he sees fit! Errors nicely updated as well! Approving for docs!

florian-lefebvre commented on packages/astro/src/actions/integration.ts:
```diff
@@ -0,0 +1,52 @@
+		name: VIRTUAL_MODULE_ID,
+		hooks: {
+			async 'astro:config:setup'(params) {
+				if (!isServerLikeOutput(params.config)) {
```
I think this should occur in config:done instead since it could be updated during any integration config:setup

bholmesdev replied:
Oh good point. I guess it would error _after_ we attempt to add middleware, but I guess that's fine

ematipico commented on packages/astro/src/actions/plugins.ts:
```diff
@@ -0,0 +1,90 @@
+			const isActionsUsed = await usesActions(fs, settings.config.srcDir);
+			// Watch for the actions file to be created.
+			async function watcherCallback() {
+				if (!isActionsUsed && (await usesActions(fs, settings.config.srcDir))) {
```
When do we change the value of `isActionsUsed`? If line 65 sets the value to `true`, then here we `false` and `server.restart()` is never called again.

ematipico commented on packages/astro/src/actions/runtime/virtual/get-action.ts:
```diff
@@ -11,7 +11,7 @@ export async function getAction(
 	const pathKeys = path.replace('/_actions/', '').split('.');
 	// @ts-expect-error virtual module
-	let { server: actionLookup } = await import('astro:internal-actions');
+	let { server: actionLookup = {} } = await import('astro:internal-actions');
```
Do we have a case where `server` is `undefined`? If so, I think it should be an error instead

bholmesdev replied:
Ah, I added that from an older prototype, but it's no longer possible to get `undefined`. I can add an internal error to flag when `actionLookup` is unexpectedly undefined or some other type

ematipico commented on packages/astro/src/actions/utils.ts:
```diff
@@ -25,3 +27,46 @@ export function createCallAction(context: ActionAPIContext): APIContext['callAct
+
+let didInitLexer = false;
+
+export async function usesActions(fs: typeof fsMod, srcDir: URL) {
```
Can we add a comment that explains what this function does? The logic is really weird and I can't understand it

bholmesdev replied:
Good call. I renamed this to `isActionsFilePresent` and added an in-line comment on why we use lexer.

ematipico commented on packages/astro/src/actions/utils.ts:
```diff
@@ -25,3 +27,46 @@ export function createCallAction(context: ActionAPIContext): APIContext['callAct
+}
+
+function search(fs: typeof fsMod, srcDir: URL) {
+	const paths = [
+		'actions.mjs',
+		'actions.js',
+		'actions.mts',
+		'actions.ts',
+		'actions/index.mjs',
+		'actions/index.js',
+		'actions/index.mts',
+		'actions/index.ts',
+	].map((p) => new URL(p, srcDir));
```
Why do we use this pattern to look for an action file? This pattern is really similar to `middleware.js` - where users can have `middleware.js` or `middleware/index.js`, and more -  and in that case we use the `moduleLoader` to retrieve the file.

It's not a blocker, but I think we should agree on **one way** to do things internally. 

Using the module loader from vite would allow us also to remove the use of `es-module-lexer`

bholmesdev replied:
Discussed on discord. We discovered that we can't rely on Vite to check whether actions are used, since we need that information _before_ vite starts to inject routes. We're tabling this for now to find a better solution.

ematipico reviewed (commented):
I have some reservations on how we load the actions file, because it adds _yet another way to do things_. 

I don't want to block the PR, but I believe we should address it.

sarah11918 reviewed (approved):
One little changeset nit, otherwise both changesets are approved by docs!

kenedyolisi commented:
Actions is one of the many features I love in astro. I couldn't wait for it to be stable and the provided guide is so extensive. I looove it. Thanks to @bholmesdev for tremendous effort on this and other contributors too for this feature.

bholmesdev commented:
Ah, you made my day! Thanks @kenedyolisi
